### PR TITLE
feat: store ABI file and commit hash of each deploy

### PR DESCRIPTION
**Pull request by 0xble**
### Motivation

Help FE track ABI changes.

### Solution

-  Store the latest commit for each deployment.
-  Store the ABI file for each deployment under `deployments/abi/[CONTRACT]/v[VERSION].json`.
- Notify team whenever ABI changes between contract versions.
- Include link to Gist of diff between previous and latest version ABIs in notification.
- Requires adding GitHub token to repo as a secret to upload the Gist.
- Requires adding Discord webhook URL as a secret to send the notification.

I considered the following features but decided against them:
-  Storing .diff files for each contract version update to compare changes between the latest and previous ABIs. This felt too opinionated. For example, what if someone wanted to view changes between the latest and the third-to-last commit? Instead, I opted to store each ABI file, searchable by contract and version.

### CODE REVIEW

File: src/deploy.ts [hunk 1]
```diff
@@ -153,24 +151,23 @@ async function runDeploy(
...
+
+  // Store ABI file
+  const abiDir = `deployments/abi/${contract}`;
+  if (!fs.existsSync(abiDir)) {
+    fs.mkdirSync(abiDir, { recursive: true });
+  }
+  fs.writeFileSync(
+    `${abiDir}/v${newDeploy.version.replace(/\./g, "_")}.json`,
+    JSON.stringify(contractJson.abi, null, 2),
+  );
```
Comment (arr00) [inline src/deploy.ts:170]: This should be option. Lets add another flag like `storeAbi` for the CLI.

File: src/deploy.ts [hunk 2]
```diff
@@ -519,3 +517,16 @@ function getProjectContracts(): string[] {
...
   return deployableContracts;
 }
+
+/**
+ * Gets the latest commit hash
+ * @returns Commit hash
+ */
+function getLatestCommitHash(): string {
+  try {
+    return execSync("git rev-parse HEAD").toString().trim();
```
Comment (arr00) [inline src/deploy.ts:527]: Should work in CI but we should test before pushing a release

**Review by arr00 — CHANGES_REQUESTED** [review]
Looks good--nice move adding the `contractJson` const. We should test in CI.

File: src/deploy.ts (hunk 2 above)
Comment (0xble) [reply src/deploy.ts:553]: Will definitely test before merging

**Comment by arr00** [comment]
What is the empty `constants.ts` file?

File: src/deploy.ts [hunk 3]
```diff
@@ -153,24 +166,34 @@ async function runDeploy(
...
     .digest("hex");
+  newDeploy.commitHash = getLatestCommitHash();
+
+  if (storeAbi) {
+    // Store ABI file and generate diff
+    const abiDir = `deployments/abi/${contract}`;
+    if (!fs.existsSync(abiDir)) {
+      fs.mkdirSync(abiDir, { recursive: true });
+    }
+    const newAbiPath = `${abiDir}/v${newDeploy.version.replace(/\./g, "_")}.json`;
```
Comment (arr00) [inline src/deploy.ts:183]: How does this fit into minor version changes?

**Review by arr00 — COMMENTED** [review]
Some thoughts after a more detailed review.

File: src/deploy.ts (hunk 3 above)
Comment (0xble) [reply src/deploy.ts:183]: Ah good catch

File: src/deploy.ts [hunk 4]
```diff
@@ -519,3 +543,170 @@ function getProjectContracts(): string[] {
...
+  const abiDir = `deployments/abi/${contract}`;
+  const files = fs.readdirSync(abiDir).sort();
+
+  if (files.length < 2) {
+    console.log(`No previous version found for ${contract}. Skipping diff generation and notification.`);
+    return;
+  }
+
+  const newAbiPath = `${abiDir}/${newVersion.replace(/\./g, "_")}.json`;
+  const previousAbiPath = `${abiDir}/${files[files.length - 2]}`;
```
[the remark this answers is not in the document]
Comment (arr00) [reply src/deploy.ts:586]: https://github.com/PartyDAO/solidity-deploy/blob/a9c24693c4045b6ba9886347df1774c6dd560ec0/src/deploy.ts#L578

**Comment by 0xble** [comment]
Going to test this, will request review after confirming it works as expected

**Comment by 0xble** [comment]
Confirmed it works
zdLGlajg

**Comment by 0xble** [comment]
Tested in a separate repo, confirmation it works
zdLGlajg

File: package.json [hunk 5]
```diff
@@ -7,12 +7,17 @@
...
     "typescript": "^5.4.5",
     "yargs": "^17.7.2"
   },
   "devDependencies": {
+    "@types/diff": "^5.2.1",
     "@types/node": "^20.14.5",
+    "@types/semver": "^7.5.8",
     "@types/yargs": "^17.0.32",
+    "axios": "^1.7.4",
+    "diff": "^5.2.0",
```
Comment (arr00) [inline package.json:20]: Aren't these not dev dependencies?

**Comment by arr00** [comment]
Seems like this PR reverts changes from #6.

**Comment by 0xble** [comment]
> Seems like this PR reverts changes from #6.

I'll fix that

File: src/deploy.ts [hunk 6]
```diff
@@ -194,9 +197,73 @@ async function runDeploy(
...
+      fs.writeFileSync(newAbiPath, JSON.stringify(contractJson.abi, null, 2))
+      const abiFiles = fs.readdirSync(abiDir)
+      const versions = sortVersions(
+        abiFiles
+          .map((file) => file.match(/\d+\_\d+\_\d+/)?.[0] || '')
+          .filter(Boolean)
+      )
+
+      // Check if there's a previous version
+      if (versions.length > 0) {
```
Comment (arr00) [inline src/deploy.ts:220]: shouldn't this be > 1?

**Review by arr00 — COMMENTED** [review]
looks good. some final nits

File: src/deploy.ts [hunk 7]
```diff
@@ -534,3 +613,242 @@ function getProjectContracts(): string[] {
...
+    }
+  }
+}
+
+async function uploadToGist(
+  content: string,
+  filename: string,
+  description: string,
+  githubToken: string
+): Promise<string> {
```
[the remark this answers is not in the document]
Comment (0xble) [reply src/deploy.ts:758]: Not sure what's wrong here. Async functions always return Promises

[the remark this answers is not in the document]
Comment (arr00) [reply src/deploy.ts:758]: ah yes my bad never-mind.

**Review by arr00 — APPROVED** [review]

**Pull request merged**